django/core/handlers/asgi.py

```diff
@@ -0,0 +1,325 @@
+            self.path_info = scope['path'][len(self.script_name):]
+        else:
+            self.path_info = scope['path']
+
+        # The Django path is different from ASGI scope path args, it should
+        # combine with script name
+        if self.script_name:
+            self.path = '%s/%s' % (
+                self.script_name.rstrip('/'),
+                self.path_info.replace('/', '', 1),
+            )
+        else:
+            self.path = scope['path']
+
+        # HTTP basics
+        self.method = self.scope['method'].upper()
+        # Ensure query string is encoded correctly
+        query_string = self.scope.get('query_string', '')
+        if isinstance(query_string, bytes):
+            query_string = query_string.decode()
+        self.META = {
+            'REQUEST_METHOD': self.method,
+            'QUERY_STRING': query_string,
+            'SCRIPT_NAME': self.script_name,
+            'PATH_INFO': self.path_info,
+            # WSGI-expecting code will need these for a while
+            'wsgi.multithread': True,
+            'wsgi.multiprocess': True,
+        }
+        if self.scope.get('client'):
+            self.META['REMOTE_ADDR'] = self.scope['client'][0]
+            self.META['REMOTE_HOST'] = self.META['REMOTE_ADDR']
+            self.META['REMOTE_PORT'] = self.scope['client'][1]
+        if self.scope.get('server'):
+            self.META['SERVER_NAME'] = self.scope['server'][0]
+            self.META['SERVER_PORT'] = str(self.scope['server'][1])
+        else:
+            self.META['SERVER_NAME'] = 'unknown'
+            self.META['SERVER_PORT'] = '0'
+        # Headers go into META
+        for name, value in self.scope.get('headers', []):
+            name = name.decode('latin1')
+            if name == 'content-length':
+                corrected_name = 'CONTENT_LENGTH'
+            elif name == 'content-type':
+                corrected_name = 'CONTENT_TYPE'
+            else:
+                corrected_name = 'HTTP_%s' % name.upper().replace('-', '_')
+            # HTTP/2 say only ASCII chars are allowed in headers, but decode
+            # latin1 just in case.
+            value = value.decode('latin1')
+            if corrected_name in self.META:
+                value = self.META[corrected_name] + ',' + value
+            self.META[corrected_name] = value
+        # Pull out request encoding if we find it
+        if 'CONTENT_TYPE' in self.META:
+            self.content_type, self.content_params = cgi.parse_header(
+                self.META['CONTENT_TYPE']
+            )
+            if 'charset' in self.content_params:
+                try:
+                    codecs.lookup(self.content_params['charset'])
+                except LookupError:
+                    pass
+                else:
+                    self.encoding = self.content_params['charset']
+        else:
+            self.content_type, self.content_params = '', {}
```

Comment:
We have exactly the same (similar) code in `WSGIRequest.__init__()`. Maybe we can add an internal hook to the `HttpRequest`? (in a separate commit) e.g.
```python
def _parse_content_type_params(self, meta):
    self.content_type, self.content_params = cgi.parse_header(meta.get('CONTENT_TYPE', ''))
    if 'charset' in self.content_params:
        try:
            codecs.lookup(self.content_params['charset'])
        except LookupError:
            pass
        else:
            self.encoding = self.content_params['charset']
```
after that we could just call:
```python
self._parse_content_type_params(self.META)
```

Let me know if it works for you, I can handle this.

Reply:
Yeah, I think we can clean this up later on. It's definitely in a couple of places.

Reply:
I added this hook in #11471.